{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Applications\n",
    "+ Applications\n",
    "+ For sequence analysis (DNA,RNA)\n",
    "+ To do transcription and translation of DNA (Protein Synthesis)\n",
    "+ Querying and Access BioInformatic Databases\n",
    "+ Entrez, BLAST,GenBank,etc\n",
    "+ 3D Structure analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Questions\n",
    "+ How to create a sequence?\n",
    "+ Protein Synthesis (Transcription,Translation)\n",
    "+ Find GC content,\n",
    "+ Molecular weights of a protein or DNA\n",
    "+ Convert protein sequence from one-letter to three-letter code and vice versa\n",
    "+ Find patterns\n",
    "+ How to find the Frequency of Nucleotide,Amino Acids,etc\n",
    "+ Sequence Alignment\n",
    "+ Similarity and homology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna,generic_rna,generic_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', DNAAlphabet())"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna = Seq('ATGATCTCGTAA',generic_dna)\n",
    "dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGAUCUCGUAA', RNAAlphabet())"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna = Seq('AUGAUCUCGUAA',generic_rna)\n",
    "rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Our Bioinformatic Pkgs\n",
    "import Bio\n",
    "\n",
    "# Methods and Attributes\n",
    "\n",
    "# dir(Bio)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Analysis\n",
    "+ DNA and RNA Sequence\n",
    "    - A Adenine\n",
    "    - C Cytosine\n",
    "    - G Guanine\n",
    "    - T Thymine\n",
    "    - U Uracil * RNA\n",
    "+ DNA Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Methods of Seq\n",
    "\n",
    "# dir(Seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create A simple sequences\n",
    "seq1 = Seq('ATGATCTCGTAA')\n",
    "seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Alphabet()"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking for the type of sequences\n",
    "seq1.alphabet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alphabet Types\n",
    "+ generic_dna/rna\n",
    "+ generic_protein\n",
    "+ IUPACUnambiguousDNA, which provides for just the basic letters, \n",
    "+ IUPACAmbiguousDNA ,which provides for ambiguity letters for every possible situation\n",
    "\n",
    "#### Usefulness of Specifying the Type of Sequence or Alphabet\n",
    "+ Help us to have an idea of the type of information the Seq object contains. \n",
    "+ Act as a means of constraining the information,\n",
    "+ As a means of type checking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a generic dna\n",
    "from Bio.Alphabet import generic_dna,generic_rna,generic_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', DNAAlphabet())"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create our DNA\n",
    "dna = Seq('ATGATCTCGTAA',generic_dna)\n",
    "dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNAAlphabet()"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the type of sequence\n",
    "dna.alphabet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGUUCUCGUAA', RNAAlphabet())"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an RNA\n",
    "rna = Seq('AUGUUCUCGUAA',generic_rna)\n",
    "rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating an IUPAC dna\n",
    "from Bio.Alphabet import IUPAC   # Standard nucleotide and protein alphabets defined by IUPAC."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    " + IUPACProtein\n",
    "    - IUPAC protein alphabet of the 20 standard amino acids.\n",
    " + IUPACAmbiguousDNA\n",
    "    - Uppercase IUPAC ambiguous DNA.\n",
    " + IUPACUnambiguousDNA\n",
    "    - Uppercase IUPAC unambiguous DNA (letters GATC only).\n",
    " + ExtendedIUPACDNA\n",
    "    - Extended IUPAC DNA alphabet.\n",
    " + IUPACAmbiguousRNA\n",
    "    - Uppercase IUPAC ambiguous RNA.\n",
    " + IUPACUnambiguousRNA\n",
    "    - Uppercase IUPAC unambiguous RNA (letters GAUC only)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_iupac = Seq('ATGATCTCGTAA',IUPAC.unambiguous_dna)\n",
    "dna_iupac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIT', ProteinAlphabet())"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating a Protein seq\n",
    "protein_a = Seq('MIT',generic_protein)\n",
    "protein_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ProteinAlphabet()"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_a.alphabet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Manipulation\n",
    "+ Indexing/Slicing\n",
    "+ Join 2 Sequences\n",
    "+ Find a Codon in a sequence\n",
    "+ Count the number of Nucleotides\n",
    "+ Find the number of times a patten repeat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', DNAAlphabet())"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12 = 12\n"
     ]
    }
   ],
   "source": [
    "# Length of our seq\n",
    "print(str(len(dna)) + \" = \" + str(len('ATGATCTCGTAA')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGAT', DNAAlphabet())"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Slicing\n",
    "dna[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AATGCTCTAGTA', DNAAlphabet())"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse\n",
    "dna[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq2 = Seq('ATGATCTCGTGG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCATGATCTCGTGG')"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join Seq\n",
    "dna[0:6] + seq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "# Find the number of Nucleotides/Base within our seq\n",
    "# Count\n",
    "print(dna.count('G'))\n",
    "print(dna.count('T'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of codon (3B) within a seq\n",
    "dna.count('ATG') # dna = ATGATCTCGTAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the position/location of a nucleotide\n",
    "dna.find('G') # dna = ATGATCTCGTAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the position/location of a nucleotide from the right\n",
    "dna.rfind('G') # dna = ATGATCTCGTAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the location/index using index\n",
    "dna.index('G')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the location/index using index\n",
    "dna.rindex('G')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot of Frequency of Nucleotides/Base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', DNAAlphabet())"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 4, 'T': 4, 'G': 2, 'C': 2})"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_freq = Counter(dna)\n",
    "dna_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 4 artists>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAO9ElEQVR4nO3dbYxcZ3nG8f9VxwQqEJHiFUR+ydKSqoIIAmwDKJVIA0jhRTGIINlVgSDoSogIkJBowodQ8gm+kAqCiEwT4VBKghKKFjBFQSGCSMVk7ToBx1C5vChuUmWJwSECQo3ufthxtZ3MeM54ZzPeJ/+fNPI557nnzO0j6/KzZ8+Zk6pCkrT+/dG0G5AkTYaBLkmNMNAlqREGuiQ1wkCXpEacMa0P3rRpU83Ozk7r4yVpXdq3b98vqmpm0NjUAn12dpbFxcVpfbwkrUtJfj5szFMuktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqRGdAz3JhiT/nuRrA8bOTHJrksNJ9iaZnWSTkqTRxpmhvx84NGTsXcAvq+r5wHXAx1fbmCRpPJ0CPckW4A3APw4p2Q7s7i3fBrw6SVbfniSpq653iv4D8CHgWUPGNwMPAFTV8STHgLOBX6wsSjIPzANs27btVPoFYPaqr5/ye1vws4+9YdX78Biu/hhKp5uRM/QkbwQerqp9JysbsO0Jj0Kqql1VNVdVczMzA7+KQJJ0irqccrkIuCzJz4BbgEuS/FNfzRFgK0CSM4BnA0cn2KckaYSRgV5VV1fVlqqaBXYAd1bV3/SVLQDv6C1f3qvxYaWS9CQ65W9bTHItsFhVC8CNwOeTHGZ5Zr5jQv1JkjoaK9Cr6i7grt7yNSu2/w546yQbkySNxztFJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmN6PKQ6Kcn+X6Se5McTPLRATVXJFlKcqD3evfatCtJGqbLE4seBy6pqseSbATuTvKNqvpeX92tVXXl5FuUJHUxMtB7D3t+rLe6sffyAdCSdJrpdA49yYYkB4CHgTuqau+AsrckuS/JbUm2TrRLSdJInQK9qv5QVRcAW4ALk5zfV/JVYLaqXgR8C9g9aD9J5pMsJllcWlpaTd+SpD5jXeVSVb8C7gIu7dv+SFU93lv9LPCyIe/fVVVzVTU3MzNzCu1KkobpcpXLTJKzesvPAF4D/Kiv5pwVq5cBhybZpCRptC5XuZwD7E6ygeX/AL5UVV9Lci2wWFULwPuSXAYcB44CV6xVw5Kkwbpc5XIf8JIB269ZsXw1cPVkW5MkjcM7RSWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRXZ4p+vQk309yb5KDST46oObMJLcmOZxkb5LZtWhWkjRclxn648AlVfVi4ALg0iSv6Kt5F/DLqno+cB3w8cm2KUkaZWSg17LHeqsbe6/qK9sO7O4t3wa8Okkm1qUkaaRO59CTbEhyAHgYuKOq9vaVbAYeAKiq48Ax4OwB+5lPsphkcWlpaXWdS5L+n06BXlV/qKoLgC3AhUnO7ysZNBvvn8VTVbuqaq6q5mZmZsbvVpI01FhXuVTVr4C7gEv7ho4AWwGSnAE8Gzg6gf4kSR11ucplJslZveVnAK8BftRXtgC8o7d8OXBnVT1hhi5JWjtndKg5B9idZAPL/wF8qaq+luRaYLGqFoAbgc8nOczyzHzHmnUsSRpoZKBX1X3ASwZsv2bF8u+At062NUnSOLxTVJIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhrR5ZmiW5N8O8mhJAeTvH9AzcVJjiU50HtdM2hfkqS10+WZoseBD1bV/iTPAvYluaOq7u+r+25VvXHyLUqSuhg5Q6+qh6pqf2/518AhYPNaNyZJGs9Y59CTzLL8wOi9A4ZfmeTeJN9I8sIh759PsphkcWlpaexmJUnDdQ70JM8Ebgc+UFWP9g3vB86tqhcDnwK+MmgfVbWrquaqam5mZuZUe5YkDdAp0JNsZDnMv1BVX+4fr6pHq+qx3vIeYGOSTRPtVJJ0Ul2ucglwI3Coqj4xpOa5vTqSXNjb7yOTbFSSdHJdrnK5CHgb8IMkB3rbPgxsA6iqG4DLgfckOQ78FthRVbUG/UqShhgZ6FV1N5ARNdcD10+qKUnS+LxTVJIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhrR5ZmiW5N8O8mhJAeTvH9ATZJ8MsnhJPcleenatCtJGqbLM0WPAx+sqv1JngXsS3JHVd2/ouZ1wHm918uBz/T+lCQ9SUbO0Kvqoara31v+NXAI2NxXth24uZZ9DzgryTkT71aSNFSXGfr/STILvATY2ze0GXhgxfqR3raH+t4/D8wDbNu2bbxOpdPI7FVfn3YLU/Wzj71hVe/3+K3u+A3T+ZeiSZ4J3A58oKoe7R8e8JZ6woaqXVU1V1VzMzMz43UqSTqpToGeZCPLYf6FqvrygJIjwNYV61uAB1ffniSpqy5XuQS4EThUVZ8YUrYAvL13tcsrgGNV9dCQWknSGuhyDv0i4G3AD5Ic6G37MLANoKpuAPYArwcOA78B3jn5ViVJJzMy0KvqbgafI19ZU8B7J9WUJGl83ikqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjejyTNGbkjyc5IdDxi9OcizJgd7rmsm3KUkapcszRT8HXA/cfJKa71bVGyfSkSTplIycoVfVd4CjT0IvkqRVmNQ59FcmuTfJN5K8cFhRkvkki0kWl5aWJvTRkiSYTKDvB86tqhcDnwK+MqywqnZV1VxVzc3MzEzgoyVJJ6w60Kvq0ap6rLe8B9iYZNOqO5MkjWXVgZ7kuUnSW76wt89HVrtfSdJ4Rl7lkuSLwMXApiRHgI8AGwGq6gbgcuA9SY4DvwV2VFWtWceSpIFGBnpV7Rwxfj3LlzVKkqbIO0UlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpESMDPclNSR5O8sMh40nyySSHk9yX5KWTb1OSNEqXGfrngEtPMv464Lzeax74zOrbkiSNa2SgV9V3gKMnKdkO3FzLvgecleScSTUoSepmEufQNwMPrFg/0tv2BEnmkywmWVxaWprAR0uSTphEoGfAthpUWFW7qmququZmZmYm8NGSpBMmEehHgK0r1rcAD05gv5KkMUwi0BeAt/eudnkFcKyqHprAfiVJYzhjVEGSLwIXA5uSHAE+AmwEqKobgD3A64HDwG+Ad65Vs5Kk4UYGelXtHDFewHsn1pEk6ZR4p6gkNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1olOgJ7k0yY+THE5y1YDxK5IsJTnQe7178q1Kkk6myzNFNwCfBl4LHAHuSbJQVff3ld5aVVeuQY+SpA66zNAvBA5X1U+q6vfALcD2tW1LkjSuLoG+GXhgxfqR3rZ+b0lyX5LbkmwdtKMk80kWkywuLS2dQruSpGG6BHoGbKu+9a8Cs1X1IuBbwO5BO6qqXVU1V1VzMzMz43UqSTqpLoF+BFg5494CPLiyoKoeqarHe6ufBV42mfYkSV11CfR7gPOSPC/J04AdwMLKgiTnrFi9DDg0uRYlSV2MvMqlqo4nuRL4JrABuKmqDia5FlisqgXgfUkuA44DR4Er1rBnSdIAIwMdoKr2AHv6tl2zYvlq4OrJtiZJGod3ikpSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjOgV6kkuT/DjJ4SRXDRg/M8mtvfG9SWYn3agk6eRGBnqSDcCngdcBLwB2JnlBX9m7gF9W1fOB64CPT7pRSdLJdZmhXwgcrqqfVNXvgVuA7X0124HdveXbgFcnyeTalCSN0uUh0ZuBB1asHwFePqymqo4nOQacDfxiZVGSeWC+t/pYkh+fStOngU30/d2eTGnj5x+P4ep4/FZnPR+/c4cNdAn0QTPtOoUaqmoXsKvDZ57WkixW1dy0+1jPPIar4/FbnVaPX5dTLkeArSvWtwAPDqtJcgbwbODoJBqUJHXTJdDvAc5L8rwkTwN2AAt9NQvAO3rLlwN3VtUTZuiSpLUz8pRL75z4lcA3gQ3ATVV1MMm1wGJVLQA3Ap9PcpjlmfmOtWz6NLDuTxudBjyGq+PxW50mj1+cSEtSG7xTVJIaYaBLUiMM9DEleXOSSvLn0+5lPUlydpIDvdd/J/mvFetPm3Z/60GS5yT55yQ/SbIvyb8lefO0+1pPkjw3yS1J/jPJ/Un2JPmzafc1KQb6+HYCd9P+L34nqqoeqaoLquoC4AbguhPrvTuQdRK9O6+/Anynqv6kql7G8r/BLdPtbP3oHcN/Ae6qqj+tqhcAHwaeM93OJsdAH0OSZwIXsfzdNQa6nkyXAL+vqhtObKiqn1fVp6bY03rzV8D/9B3DA1X13Sn2NFEG+njeBPxrVf0HcDTJS6fdkJ4yXgjsn3YT69z5wL5pN7GWDPTx7GT5y8no/blzir3oKSzJp5Pcm+Seafei00eX73IRy7/UY/nH3vOTFMs3WVWSD3lXrJ4EB4G3nFipqvcm2QQsTq+ldecgy3eyN8sZeneXAzdX1blVNVtVW4GfAn855b701HAn8PQk71mx7Y+n1cw6dSdwZpK/PbEhyV8kedUUe5ooA727nSz/hnyl24G/nkIveorp/RT4JuBVSX6a5PssP4Pg76bb2frRO4ZvBl7bu2zxIPD3PPHLBtctb/2XpEY4Q5ekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqRH/Czrh4nBuz6SLAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.bar(dna_freq.keys(),dna_freq.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein Synthesis\n",
    "+ Protein synthesis is how cells make proteins using 2 stages\n",
    " - Transcription\n",
    " - Translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "dna_seq = Seq('ATGATCTCGTAA')\n",
    "len(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "## A T = 2 hydrogen bonds\n",
    "## G C =  3 hydrogen bonds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACTAGAGCATT')"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Complement\n",
    "dna_seq.complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TTACGAGATCAT')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reverse complement\n",
    "dna_seq.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TTACGAGATCAT')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# custom fxn\n",
    "dna_seq.complement()[::-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Synthesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGAUCUCGUAA', RNAAlphabet())"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transcription\n",
    "# DNA to mRNA\n",
    "# Writing the msg\n",
    "dna_seq.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA = dna_seq.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translation \n",
    "# mRNA to Protein/Amino Acid\n",
    "# Method 1\n",
    "mRNA.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 2\n",
    "# Direct translation of DNA to Amino Acid\n",
    "dna_seq.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS@', HasStopCodon(ExtendedIUPACProtein(), '@'))"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create our custom stop codon sym\n",
    "mRNA.translate(stop_symbol=\"@\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA', DNAAlphabet())"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Back Transcription \n",
    "# mrna to DNA\n",
    "mRNA.back_transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mRNA.back_transcribe() == dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join the steps\n",
    "dna_seq.transcribe().translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert Amino Acid to 3 letter\n",
    "from Bio.SeqUtils import seq3,seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa = dna_seq.translate()\n",
    "aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MetIleSerTer'"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq3(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MIS*'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From 3 letters to 1\n",
    "seq1('MetIleSerTer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino Acids\n",
    "![](./images/aminoacidchart1.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# View our codonTable\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "# methods \n",
    "# dir(CodonTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.unambiguous_dna_by_name['Standard'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  U      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "U | UUU F   | UCU S   | UAU Y   | UGU C   | U\n",
      "U | UUC F   | UCC S   | UAC Y   | UGC C   | C\n",
      "U | UUA L   | UCA S   | UAA Stop| UGA Stop| A\n",
      "U | UUG L(s)| UCG S   | UAG Stop| UGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CUU L   | CCU P   | CAU H   | CGU R   | U\n",
      "C | CUC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CUA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CUG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | AUU I   | ACU T   | AAU N   | AGU S   | U\n",
      "A | AUC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | AUA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | AUG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GUU V   | GCU A   | GAU D   | GGU G   | U\n",
      "G | GUC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GUA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GUG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "# Rna codon table\n",
    "print(CodonTable.unambiguous_rna_by_name['Standard'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Composition : GC and AT Content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GC Contents In DNA\n",
    "+ GC-content (or guanine-cytosine content) is the percentage of nitrogenous bases in a DNA or RNA molecule \n",
    "that are either Guanine (G) or Cytosine (C)\n",
    "\n",
    "A=T \n",
    "G=-C\n",
    "\n",
    "#### Usefulness\n",
    "+ In polymerase chain reaction (PCR) experiments, the GC-content of short oligonucleotides known as primers is often used to predict their annealing temperature to the template DNA. \n",
    "+ A higher GC-content level indicates a relatively higher melting temperature.\n",
    "+ DNA with low GC-content is less stable than DNA with high GC-content\n",
    "+ High GC content DNA can make it difficult to perform PCR amplication due to difficulty in designing a primer long enough to provide great specifity\n",
    "\n",
    "\n",
    "#### AT Contents in DNA\n",
    "+ AT content is the percentage of nitrogenous bases in a DNA or RNA molecule that are either Adenine (A) or Thymine (T)\n",
    "+ AT base pairing yields only 2 hydrogen bonds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import GC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.333333333333336"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 2\n",
    "# Custom Fxn to Get GC Count\n",
    "dna_seq.count('A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.33333333333333"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def gc_content(seq):\n",
    "    result = float(seq.count('G') + seq.count('C')) / len(seq) * 100\n",
    "    return result\n",
    "\n",
    "gc_content(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.33333333333333"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 3\n",
    "def gc_content2(seq):\n",
    "    gc = [Base for Base in seq.upper() if Base in 'GC']\n",
    "    result = float(len(gc))/len(seq) * 100\n",
    "    return result\n",
    "\n",
    "gc_content2(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('atgatctcgtaa')"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.33333333333333"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_content2('atgatctcgtaa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.333333333333336"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC('atgatctcgtaa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AT Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fxn for AT Content\n",
    "\n",
    "def at_content(seq):\n",
    "    result = float(seq.count('A') + seq.count('T')) / len(seq) * 100\n",
    "    return result\n",
    "\n",
    "at_content(dna_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Melting Point of DNA\n",
    "\n",
    "+ Higher GC means high melting point\n",
    "+ Tm_Wallace: 'Rule of thumb'\n",
    "+ Tm_GC: Empirical formulas based on GC content. Salt and mismatch corrections can be included.\n",
    "+ Tm_NN: Calculation based on nearest neighbor thermodynamics. Several tables for DNA/DNA, DNA/RNA and RNA/RNA hybridizations are included. Correction for mismatches, dangling ends, salt concentration and other additives are available.\n",
    "    - further info : https://biopython.org/DIST/docs/api/Bio.SeqUtils.MeltingTemp-module.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio.SeqUtils\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "# dir(Bio.SeqUtils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.333333333333336"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wallence : 32.0 and GC : 23.32155893208184 \n"
     ]
    }
   ],
   "source": [
    "# Check for the melting point using wallace\n",
    "w  = mt.Tm_Wallace(dna_seq)\n",
    "\n",
    "# Check for the melting point using GC content\n",
    "gc = mt.Tm_GC(dna_seq)\n",
    "\n",
    "print(\"Wallence : {} and GC : {} \".format(w,gc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise\n",
    "+ Which of the following sequence will have the highest GC ?\n",
    "+ ex1 = 'ATGCATGGTGCGCGA'\n",
    "+ ex2 = 'ATTTGTGCTCCTGGA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "ex1 = 'ATGCATGGTGCGCGA'\n",
    "ex2 = 'ATTTGTGCTCCTGGA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_metrics(seq):\n",
    "    gc = GC(seq)\n",
    "    at = at_content(seq)\n",
    "    melting = mt.Tm_GC(seq)\n",
    "    result = \"GC : {}, AT : {}, Temp : {}\".format(gc,at,melting)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC : 60.0, AT : 40.0, Temp : 44.254892265415165\n",
      "GC : 46.666666666666664, AT : 53.333333333333336, Temp : 38.7882255987485\n"
     ]
    }
   ],
   "source": [
    "print(get_metrics(ex1))\n",
    "print(get_metrics(ex2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GC Skew\n",
    "+ check when the nucleotide (G,C) are over or under abundant in a particular region of a DNA or RNA\n",
    "+ Helps to indictate DNA lagging strand or leading strand\n",
    "+ GC skew pos = leading\n",
    "+ GC skew neg = lagging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.SeqUtils import GC123, GC_skew, xGC_skew\n",
    "\n",
    "# further info : https://biopython.org/DIST/docs/api/Bio.SeqUtils-module.html\n",
    "\n",
    "\n",
    "# GC(seq)\n",
    "# Calculate G+C content, return percentage (as float between 0 and 100).\n",
    "\n",
    "# GC123(seq)\n",
    "# Calculate G+C content: total, for first, second and third positions.\n",
    "\n",
    "# GC_skew(seq, window=100)\n",
    "# Calculate GC skew (G-C)/(G+C) for multiple windows along the sequence.\n",
    "\n",
    "# xGC_skew(seq, window=1000, zoom=100, r=300, px=100, py=100)\n",
    "# Calculate and plot normal and accumulated GC skew (GRAPHICS !!!).\n",
    "\n",
    "\n",
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.333333333333336"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33.333333333333336, 0.0, 25.0, 75.0)"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC123(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0, 0.0]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GC_Skew\n",
    "GC_skew(dna_seq,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GC_skew('ATGGGGTCCCGCTC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsequences\n",
    "+ Search for a DNA subseq in sequence, return list of [subseq, positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import nt_search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATT', 3]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_seq = Seq('ACTATT')\n",
    "sub_seq  = Seq('ATT')\n",
    "\n",
    "nt_search(str(main_seq), str(sub_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence Alignment\n",
    "+ Sequence alignment is a method of arranging sequences of DNA, RNA, or Amino Acids or proteins to identify regions of similarity. \n",
    "+ The similarity being identified, may be a result of functional, structural, or evolutionary relationships between the sequences.\n",
    "+ It is useful in identifying similarity and homology\n",
    "+ Homology: descent from a common ancestor or source.\n",
    "\n",
    "#### Terms\n",
    "+ Matches\n",
    "+ Mismatches\n",
    "+ Gap\n",
    "\n",
    "![](./images/aligngap.png)\n",
    "\n",
    "#### Alignment Types\n",
    "+ Global alignment: finds the best concordance/agreement betwenn all characters in two sequences\n",
    "    + Mostly from end to end\n",
    "    + By Needle\n",
    "+ Local Alignment: finds just the subsequences that align the best\n",
    "    + In this method, we consider subsequences within each of the 2 sequences and try to match them to obtain the best alignment.\n",
    "    + By Water\n",
    " \n",
    "![](./images/localvsglobalalignment.jpg)\n",
    "\n",
    "#### When to use local alignment\n",
    "\n",
    "+ 2 sequences have a small matched region\n",
    "+ 2 Sequences are of different lengths\n",
    "+ Overlapping sequences\n",
    "+ One sequences is a subsequences of the other\n",
    "+ Blast\n",
    "+ Emboss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "seq1 = Seq('ACTCGT')\n",
    "seq2 = Seq('ATTCG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global Alignment\n",
    "alignments = pairwise2.align.globalxx(seq1, seq2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Alignment(seqA='ACT-CGT', seqB='A-TTCG-', score=4.0, start=0, end=7),\n",
       " Alignment(seqA='AC-TCGT', seqB='A-TTCG-', score=4.0, start=0, end=7),\n",
       " Alignment(seqA='ACTCGT', seqB='ATTCG-', score=4.0, start=0, end=6)]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACT-CGT\n",
      "| | || \n",
      "A-TTCG-\n",
      "  Score=4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# To display the alignment\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACT-CGT\n",
      "| | || \n",
      "A-TTCG-\n",
      "  Score=4\n",
      "\n",
      "AC-TCGT\n",
      "|  ||| \n",
      "A-TTCG-\n",
      "  Score=4\n",
      "\n",
      "ACTCGT\n",
      "|.||| \n",
      "ATTCG-\n",
      "  Score=4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# View all\n",
    "for i in alignments:\n",
    "    print(format_alignment(*i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Local Alignment\n",
    "loc_alignments = pairwise2.align.localxx(seq1, seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 ACT-CG\n",
      "  | | ||\n",
      "1 A-TTCG\n",
      "  Score=4\n",
      "\n",
      "1 AC-TCG\n",
      "  |  |||\n",
      "1 A-TTCG\n",
      "  Score=4\n",
      "\n",
      "1 ACTCG\n",
      "  |.|||\n",
      "1 ATTCG\n",
      "  Score=4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# View all\n",
    "for i in loc_alignments:\n",
    "    print(format_alignment(*i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.0"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the aligment by only the score\n",
    "alignments2 = pairwise2.align.globalxx(seq1,seq2,one_alignment_only=True,score_only=True)\n",
    "alignments2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "float"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(alignments2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check for similarity or percentage of similarity using Alignment\n",
    "+ fraction of nucleotides that is the same/ total number of nucleotides * 100%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ACTCGT')"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTCG')"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments2/len(seq1) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the alignment by only the score\n",
    "loc_alignment2 = pairwise2.align.localxx(seq1,seq2,one_alignment_only=True,score_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loc_alignment2/len(seq1) * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find out all the possible global alignments with the maximum similarity score\n",
    "+ Matching characters :2 points, \n",
    "+ Each mismatching character: -1 point\n",
    "+ 0.5 points are deducted when opening a gap, \n",
    "+ 0.1 points are deducted when extending it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GLobal alignment with max sim\n",
    "glb_alignment = pairwise2.align.globalms(seq1,seq2,2,-1,-0.5,-0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACT-CGT\n",
      "| | || \n",
      "A-TTCG-\n",
      "  Score=6.5\n",
      "\n",
      "AC-TCGT\n",
      "|  ||| \n",
      "A-TTCG-\n",
      "  Score=6.5\n",
      "\n",
      "ACTCGT\n",
      "|.||| \n",
      "ATTCG-\n",
      "  Score=6.5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# View all\n",
    "for i in glb_alignment:\n",
    "    print(format_alignment(*i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A v B :  100.0\n",
      "B v C :  100.0\n",
      "A v C :  85.71428571428571\n"
     ]
    }
   ],
   "source": [
    "seqA = Seq('AAGGCTT')\n",
    "seqB = Seq('AAGGC')\n",
    "seqC = Seq('AAGGCAT')\n",
    "AvB = pairwise2.align.localxx(seqA,seqB,one_alignment_only=True,score_only=True)\n",
    "BvC = pairwise2.align.localxx(seqB,seqC,one_alignment_only=True,score_only=True)\n",
    "AvC = pairwise2.align.localxx(seqA,seqC,one_alignment_only=True,score_only=True)\n",
    "print(\"A v B : \", AvB/len(seqB) *100)\n",
    "print(\"B v C : \", BvC/len(seqB) *100)\n",
    "print(\"A v C : \", AvC/len(seqA) *100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking for Similarity Between Sequences\n",
    "+ Sequence Alignment\n",
    "    - Dynamic Programming (Global/Local/(needle/water))\n",
    "    - Dotplot\n",
    "    \n",
    "+ Similarity: resemblance between two sequences in comparison\n",
    "    - the minimal number of edit operations (inserts, deletes, and substitutions) in order to transform the one sequence into an exact copy of the other sequence being aligned \n",
    "    - distance\n",
    "+ Identity: the number of charaters that match EXACTLY between two different sequences\n",
    "    + Gaps are not counted \n",
    "    + The measurement is relational to the shorter of the two sequences. \n",
    "    + This has the effect that sequence identity is not transitive, i.e. \n",
    "    + if sequence A=B and B=C then A is not necessarily equal C (in terms of the identity distance measure) :\n",
    " \n",
    " \n",
    " - A: AAGGCTT\n",
    " - B: AAGGC\n",
    " - C: AAGGCAT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ Here identity(A,B)=100% (5 identical nucleotides / min(length(A),length(B))).\n",
    "Identity(B,C)=100%, but identity(A,C)=85% ((6 identical nucleotides / 7)). So 100% identity does not mean two sequences are the same.\n",
    "+ Sequence similarity is first of all a general description of a relationship but nevertheless its more or less common practice to define similarity as an optimal matching problem (for sequence alignments or unless defined otherwise). \n",
    "+ Hereby, the optimal matching algorithm finds the minimal number of edit operations (inserts, deletes, and substitutions) in order to transform the one sequence into an exact copy of the other sequence being aligned (edit distance). \n",
    "+ Using this, the percentage sequence similarity of the examples above are sim(A,B)=60%, sim(B,C)=60%, sim(A,C)=86% (semi-global, sim=1-(edit distance/unaligned length of the shorter sequence)). But there are other ways to define similarity between two objects (e.g. using tertiary strucure of proteins).\n",
    "An then you might start to conclude from similarity to homology, but this was already covered sufficiently\n",
    "+ read more https://www.researchgate.net/post/Homology_similarity_and_identity-can_anyone_help_with_these_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AvB 100.0\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import pairwise2\n",
    "\n",
    "seqA = Seq('AAGGCTT')\n",
    "seqB = Seq('AAGGC')\n",
    "seqC = Seq('AAGGCAT')\n",
    "\n",
    "AvB = pairwise2.align.localxx(seqA,seqB,one_alignment_only=True,score_only=True)\n",
    "BvC = pairwise2.align.localxx(seqB,seqC,one_alignment_only=True,score_only=True)\n",
    "AvC = pairwise2.align.localxx(seqA,seqC,one_alignment_only=True,score_only=True)\n",
    "\n",
    "print(\"AvB\",AvB/len(seqB) *100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Narative\n",
    "+ Seq A and Seq B are 100 identical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BvC :  100.0\n",
      "AvC :  85.71428571428571\n"
     ]
    }
   ],
   "source": [
    "print(\"BvC : \",BvC/len(seqB) *100)\n",
    "print(\"AvC : \",AvC/len(seqC) *100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if same\n",
    "seqA == seqB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Hamming distance: shows how many places 2 strings differ\n",
    "+ Hamming distance between two strings of equal length is the number of positions at which the corresponding symbols are different. \n",
    "+ In other words, it measures the minimum number of substitutions required to change one string into the other, or the minimum number of errors that could have transformed one string into the othe\n",
    "+ It is used for error detection or error correction\n",
    "+ It is used to quantify the similarity of DNA sequences,\n",
    "+ For checking the edit distance\n",
    " - edit distance is a way of quantifying how dissimilar two strings (e.g., words) are to one another by counting the minimum number of operations required to transform one string into the other. \n",
    " - eg Levenshtein distance\n",
    "\n",
    "#### Levenshtein distance\n",
    "+ This method was invented in 1965 by the Russian Mathematician Vladimir Levenshtein (1935-2017). \n",
    "+ The distance value describes the minimal number of deletions, insertions, or substitutions that are required to transform one string (the source) into another (the target). \n",
    "+ Unlike the Hamming distance, the Levenshtein distance works on strings with an unequal length\n",
    "\n",
    "+ pip install python-Levenshtein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = Seq('ACTAT')\n",
    "seq2 = Seq('ACTTA')\n",
    "seq3 = Seq('ACTT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Hamming Distance fxn\n",
    "def hamming_distance(lhs,rhs):\n",
    "    return len([(x,y) for x,y in zip(lhs,rhs) if x != y])\n",
    "\n",
    "hamming_distance(seq1, seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 0 if the same\n",
    "hamming_distance(seq1, seq1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTAT\n",
      "TATCA\n"
     ]
    }
   ],
   "source": [
    "print(seq1)\n",
    "print(seq1[::-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hamming_distance(seq1,seq1[::-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hamming_distance(seq1,seq3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Levenshtein Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install python-Levenshtein\n",
    "from Levenshtein import distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distance(str(seq1),str(seq2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hamming Distance 2\n",
      "Levenshtein Distance 2\n"
     ]
    }
   ],
   "source": [
    "print(\"Hamming Distance\",hamming_distance(seq1,seq2))\n",
    "print(\"Levenshtein Distance\",distance(str(seq1),str(seq2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hamming Distance 1\n",
      "Levenshtein Distance 1\n"
     ]
    }
   ],
   "source": [
    "print(\"Hamming Distance\",hamming_distance(seq1,seq3))\n",
    "print(\"Levenshtein Distance\",distance(str(seq1),str(seq3)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dot Plot\n",
    "+ A dot plot is a graphical method that allows the comparison of two biological sequences \n",
    "and identify regions of close similarity between them.\n",
    "+ Simplest method - put a dot wherever\n",
    "sequences are identical \n",
    "+ Dot plots compare two sequences by organizing one sequence on the x-axis, and another on the y-axis, of a plot. \n",
    "+ When the residues of both sequences match at the same location on the plot, a dot is drawn at the corresponding position\n",
    "\n",
    "#### usefulness\n",
    "+ Dot plots can also be used to visually inspect sequences for \n",
    "  - direct or inverted repeats\n",
    "  - regions with low sequence complexity.\n",
    "  - Similar regions\n",
    "  - Repeated sequences\n",
    "  - Sequence rearrangements\n",
    "  - RNA structures\n",
    "  - Gene order\n",
    "\n",
    "+ Link :https://stackoverflow.com/questions/40822400/how-to-create-a-dotplot-of-two-dna-sequence-in-python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def delta(x,y):\n",
    "    return 0 if x == y else 1\n",
    "\n",
    "\n",
    "def M(seq1,seq2,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip(seq1[i:i+k],seq2[j:j+k]))\n",
    "\n",
    "\n",
    "def makeMatrix(seq1,seq2,k):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    return [[M(seq1,seq2,i,j,k) for j in range(m-k+1)] for i in range(n-k+1)]\n",
    "\n",
    "\n",
    "def plotMatrix(M,t, seq1, seq2, nonblank = chr(0x25A0), blank = ' '):\n",
    "    print(' |' + seq2)\n",
    "    print('-'*(2 + len(seq2)))\n",
    "    for label,row in zip(seq1,M):\n",
    "        line = ''.join(nonblank if s < t else blank for s in row)\n",
    "        print(label + '|' + line)\n",
    "\n",
    "\n",
    "def dotplot(seq1,seq2,k = 1,t = 1):\n",
    "    M = makeMatrix(seq1,seq2,k)\n",
    "    plotMatrix(M, t, seq1,seq2) #experiment with character choice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " |AC\n",
      "----\n",
      "A|■ \n",
      "C| ■\n",
      "T|  \n",
      "T|  \n",
      "A|■ \n",
      "G|  \n"
     ]
    }
   ],
   "source": [
    "seq1 = Seq('ACTTAG')\n",
    "seq2 = Seq('AC')\n",
    "\n",
    "dotplot(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " |ACTTAG\n",
      "--------\n",
      "A|■   ■ \n",
      "C| ■    \n",
      "T|  ■■  \n",
      "T|  ■■  \n",
      "A|■   ■ \n",
      "G|     ■\n"
     ]
    }
   ],
   "source": [
    "# Indentical show diagonal\n",
    "dotplot(seq1,seq1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " |LOVE\n",
      "------\n",
      "L|■   \n",
      "O| ■  \n",
      "V|  ■ \n",
      "E|   ■\n"
     ]
    }
   ],
   "source": [
    "# Indentical show diagonal\n",
    "dotplot('LOVE','LOVE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOgAAAD4CAYAAADmf6rjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAIsElEQVR4nO3dz4td9R3G8edxqhF/rMZgKpVIC6m0QUIc7ULBX4i7YmgojlsxUNxb6Gr+AFeWgmSnizIUxK7EUJTSSgvtTUkLcRGwwUXTwhhXImhIny4yg+NknDmT3ON90vt+gTDneHPyWdw335O5557jJALQ6aZZDwDg6xEoUIxAgWIEChQjUKDYt8Y68C3el1t1+1iH37NDD3w26xGudvPhWU+AGTh9+vTHSfYPee1ogd6q2/UjPzXW4ffs1Kkzsx7hKjcdmMx6BMyA7Y+GvpZTXKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFig0O1PYx27F9/5gDAfjSXlbQZUnvS3pupFkAbDEoUNt3SHpE0gsiUOAbM3QFfVbSO0nOSfrE9tHtXmT7hO2J7cklfT61IYF5NTTQZUmr6z+vrm9fJcnJJEtJlm7WvmnMB8y1XW95YntR0pOSDtuOpAVJsf1yuC09MKohK+hxSW8kOZjkviT3Sjov6dFxRwMwJNBlSW9t2fempOenPw6AzXY9xU3y+Db7Xh1lGgBfwZVEQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiu14sf60OPfBZ1WPnn7nnyKxHuMrv/jvrCdCOFRQoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUIxAgWIEChQjUKDY4EBtH7C9avtD2x/Yftv2oTGHA+bdoEBtW1ceQfj7JN9L8gNJv5B095jDAfNu6B0VnpB0KclrGzuS9NwuAfg/NfQU97Ck07u9yPYJ2xPbk7WLl69vMgDT/SVRkpNJlpIs7V9cmOahgbk0NNCzkh4ccxAAVxsa6HuS9tl+cWOH7YdsPzbOWACkgYEmiaRjkp5e/5jlrKQVSRdGnA2Ye4Pvi5vkgqSfjjgLgC24kggoRqBAMQIFihEoUIxAgWIEChQjUKAYgQLFCBQoRqBAMQIFihEoUGzwxfI3ulMXuEMLbjysoEAxAgWKEShQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoNigr5vZXpT07vrmAUmXJa2tbz+c5IsRZgPm3qBAk1yUdESSbK9I+jTJKyPOBUCc4gLVphqo7RO2J7YnaxcvT/PQwFyaaqBJTiZZSrK0f3FhmocG5hKnuEAxAgWKEShQbM+33UyyMsIcALbBCgoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1CgGIECxQY9m8X2oqR31zcPSLosaW19++EkX4wwGzD3BgWa5KKkI5Jke0XSp0leGXEuAOIUF6g21UBtn7A9sT1Zu3h5mocG5tJUA01yMslSkqX9iwvTPDQwlzjFBYoRKFCMQIFigz5m2SzJyghzANgGKyhQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoNieL5Yf6tw/btMz9xwZ6/B7durCmVmPAOwZKyhQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoBiBAsUIFCg2OFDbx2zH9v1jDgTgS3tZQZclvS/puZFmAbDFoEBt3yHpEUkviECBb8zQFfRZSe8kOSfpE9tHt3uR7RO2J7Ynl/T51IYE5tXQQJclra7/vLq+fZUkJ5MsJVm6WfumMR8w13a9J5HtRUlPSjpsO5IWJMX2y0ky9oDAPBuygh6X9EaSg0nuS3KvpPOSHh13NABDAl2W9NaWfW9Ken764wDYbNdT3CSPb7Pv1VGmAfAVXEkEFCNQoBiBAsUIFChGoEAxAgWKEShQjECBYgQKFCNQoBiBAsUIFChGoECxXb/Ncq0OPfCZTp06M9bhgbnACgoUI1CgGIECxQgUKEagQDECBYoRKFCMQIFiBAoUI1CgGIECxQgUKEagQDECBYoNCtT23bZ/bfuftk/b/rPtY2MPB8y7XQO1bUm/lfSHJN9N8qCk5yR9Z+zhgHk35AvbT0r6IslrGzuSfCTpl6NNBUDSsFPcH0r625CD2T5he2J7snbx8vVNBmDvvySy/Svbf7f9163/L8nJJEtJlvYvLkxnQmCODQn0rKSjGxtJXpL0lKT9Yw0F4Iohgb4n6VbbP9u077aR5gGwya6BJomkZyU9Zvu87b9Iel3Sz8ceDph3g267meTfuvLRCoBvEFcSAcUIFChGoEAxAgWKEShQjECBYgQKFCNQoBiBAsUIFChGoEAxAgWK+cqXVUY4sL0m6aMpHOouSR9P4TjT0jaP1DcT8+zs+0nuHPLCQd9muRZJpvKFbtuTJEvTONY0tM0j9c3EPDuzPRn6Wk5xgWIEChS7EQI9OesBtmibR+qbiXl2Nnie0X5JBOD63QgrKDC3CBQoVh2o7WO2Y/v+glkO2F61/aHtD2y/bfvQjGZZtH1m/b//2P7Xpu1bmOkrczW9h/b8ELLqf4Pa/o2kb0t6N8nKDOewpD9Jen3jGTW2j0i6M8kfZzXX+hwrkj5N8sos59isaaby99BBST9O8rXPOapdQW3fIekRSS9o9rf8fELSpS0PkDoz6zixs7L30LYPIdspTqk4UF25WfY7Sc5J+sT20d3+wIgOSzo9w78f16bpPTT4IWSbNQe6LGl1/efV9W1gL2rfQzs9hGyz0a7FvR62F3XllOCw7UhakBTbL2c2/2g+K+n4DP5eXKPS99BPNjaSvGT7Lkk7XpfbuoIel/RGkoNJ7ktyr6Tzkh6d0TzvSdpn+8WNHbYfsv3YjObB7hrfQ3t+CFlroMuS3tqy701Jz89glo0HSB2T9PT6xyxnJa1IujCLeTBI43tozw8hq/6YBZh3rSsoABEoUI1AgWIEChQjUKAYgQLFCBQo9j+xSxrC+2z4dwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Add Some fancyness to it\n",
    "plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "xt=plt.xticks(np.arange(len(list(seq1))),list(seq1))\n",
    "yt=plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAGUAAAD4CAYAAAD1hChSAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAGf0lEQVR4nO3dwYtd5R3G8e/TmE7Q7KwkSCWiMIgNIcSYLpRGEyTbDARx3AaF0r0FN50/oCulELKrizIbSekiZBYGEaHQJkWFCAY0ihLBmKwkoCH8usgZO0kmzTt43pknnuezmjl3uPcHX95z58697xxVFeHlFxs9QNwuUQwliqFEMZQohu7rdce/1Ext4YFed79ms7uubvQIN/n8y2t8e+W6VrutW5QtPMBvdbDX3a/Z0tIHGz3CTfYd+vKOt+X0ZShRDCWKoUQxlCiGEsVQohhKFEOJYihRDCWKoUQxlCiGmqNImpNUkp7oOVCsbaXMA+8DL3WaJQZNUSRtBZ4BjpIo3bWulMPAqao6D1yRtGe1H5L0qqQzks5c4/vRhpya1ijzwOLw9eLw/W2q6nhV7a2qvZuZGWO+Sbrr28GSHgQOADslFbAJKEmvVT5e2UXLSjkCvFVVO6rq0ap6BLgAPNt3tOlqiTIPnLjl2NvAy+OPE9Bw+qqq51Y59kaXaQLIK3pLiWIoUQwliqFEMZQohhLFUKIYShRDiWIoUQwliqFu2+tmd1212tJ26OHdGz3CTc7X5TvelpViKFEMJYqhRDGUKIYSxVCiGEoUQ4liKFEMJYqhRDGUKIYSxVCiGFrLRtTtkhYlfSrpY0knJc32HG6qWvc8ihvbId6tqser6kngdWBbz+GmqvWdx+eBa1V1bPlAVfm8rfgz03r62gmcvdsPrdyIeuny9Z822YSN+kS/ciPqQw9uGvOuJ6U1yjngqZ6DxP+0RjkNzEh6ZfmApKcl7e8z1rQ1RRm2Zs8BLwy/Ep8DFoCLHWebrObPfVXVReDFjrPEIK/oDSWKoUQxlCiGEsVQohhKFEOJYihRDCWKoUQxlCiGum1EdbN00evd632H7nwx0KwUQ4liKFEMJYqhRDGUKIYSxVCiGEoUQ4liKFEMJYqhRDGUKIaa/nQ/XJfrneHb7cB14NLw/b6q+qHDbJPVFKWqLgO7ASQtAN9V1Z87zjVpOX0ZGjVK9jyOI3seDeX0ZShRDCWKoTV/xKiqFjrMEStkpRhKFEOJYihRDCWKoUQxlCiGEsVQohhKFEOJYihRDCWKoUQxlCiGEsVQohhKFEOJYihRDCWKoUQxlCiGEsVQohhKFEOJYih7Hg1lz6OhnL4MZc+joex5NJTTl6FEMZQohrLn0VBWiqFEMZQohhLFUKIYShRDiWIoUQwliqFEMZQohhLFULfrPJ7/6H4OPby7192vmdt1Hv+frBRDiWIoUQwliqFEMZQohhLFUKIYShRDiWIoUQwliqFEMZQohpqjSJqTVJKe6DlQrG2lzAPvAy91miUGTVEkbQWeAY6SKN21rpTDwKmqOg9ckbRntR9auefxGt+PNuTUtEaZBxaHrxeH72+zcs/jZmbGmG+S7voe/fCPDQ4AOyUVsAkoSa9VVfUecIpaVsoR4K2q2lFVj1bVI8AF4Nm+o01XS5R54MQtx94GXh5/nICG01dVPbfKsTe6TBNAXtFbShRDiWIoUQwliqFEMZQohhLFUKIYShRDiWIoUQwliqFuG1Fnd11laene2fzpJCvFUKIYShRDiWIoUQwliqFEMZQohhLFUKIYShRDiWIoUQwliqHW7XXbJP1N0meSzkr6p6S53sNN1V2jSBLwd+C9qnqsqp7ixr7HX/cebqpa3uQ6APxQVceWD1TVF8Cb3aaauJbT12+A/7TcWS6+OY41P9FL+oukDyX9+9bbcvHNcbREOQf8uEW7qv4AHAQe6jXU1LVEOQ1skfT7Fcfu7zRP0BBl2JZ9GNgv6YKkfwF/Bf7Ye7ipar0e/dfk33+sm7yiN5QohhLFUKIYShRDiWIoUQwliqFEMZQohhLFUKIYUq//zSnpEvDFCHf1K+DbEe5nLGPNs6OqVn1PqluUsUg6U1V7N3qOZesxT05fhhLF0L0Q5fhGD3CL7vPYP6dM0b2wUiYnUQxZR3G6upGk7ZIWJX0q6WNJJyXN9ngs6yiYXN1o+JD7CeDdqnq8qp4EXge2dXk81yf64epGnwDPA/+oqg1bLZIOAAtV9bv1eDznldJ0daN1shM4u14P5hyl6epGP0eWp6/h6kZfAd8AP17diBt/xFv3gSUdBP409dOX29WNTgMzkl5ZPiDpaUn7ezyYaxSrqxsNq3MOeGH4lfgcsABc7PF4lqevqXNdKZOWKIYSxVCiGEoUQ4liKFEM/ReE/6UwAQRqIQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convert to Fxn\n",
    "def dotplotx(seq1,seq2):\n",
    "    plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "    # on x-axis list all sequences of seq 2\n",
    "    xt=plt.xticks(np.arange(len(list(seq2))),list(seq2))\n",
    "    # on y-axis list all sequences of seq 1\n",
    "    yt=plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "dotplotx(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna1 = Seq(\"ATGATCTCGTAA\",generic_dna)\n",
    "dna2 = Seq(\"ATTATGTCGTAA\",generic_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " |ATTATGTCGTAA\n",
      "--------------\n",
      "A|■  ■      ■■\n",
      "T| ■■ ■ ■  ■  \n",
      "G|     ■  ■   \n",
      "A|■  ■      ■■\n",
      "T| ■■ ■ ■  ■  \n",
      "C|       ■    \n",
      "T| ■■ ■ ■  ■  \n",
      "C|       ■    \n",
      "G|     ■  ■   \n",
      "T| ■■ ■ ■  ■  \n",
      "A|■  ■      ■■\n",
      "A|■  ■      ■■\n"
     ]
    }
   ],
   "source": [
    "dotplot(dna1,dna2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPYAAAD4CAYAAADfGBu4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAALfElEQVR4nO3d0auVVRrH8d9vnFKc6CITZUiMAomSErXTRQ2WEt56mIhOt9EwEXM3FHR1/gCvikC8q4s4DIRdRV4oUYFkGjlwuhBKJMYGHL0IR1KRZy7OPszBTrPX3nst9/KZ7weE8757s3wW7/m5Xvf7vs92RAhALr+ZdgEA6iPYQEIEG0iIYAMJEWwgod+2Gvhur411+t3E42x7/GqFauo5+/f11cbqbW4Yrubxn9TP+reuxzWv9ppbXe661/fFU9438ThHL3xToZp69v9+R7Wxepsbhqt5/Cf1ZRzTT3F51WBzKg4kRLCBhAg2kBDBBhIqDrbtWdth+5GWBQGY3Cgr9pykLyS91KgWAJUUBdv2PZKelvSKCDbQvdIV+4CkTyLirKTLtneu9ibbf7J9yvapG7pWrUgAoykN9pykhcHPC4PtX4iIwxGxOyJ236W1NeoDMIaht5Ta3iBpr6TttkPSGklh+42gSwPQpZIV+wVJ70fE1oh4MCK2SDon6Zm2pQEYV0mw5yQduWXfh5Jerl8OgBqGnopHxLOr7Hu7STUAquDOMyAhgg0kRLCBhJp1UNn2+FUdPTp5I4GeHmyX6jZHqDW3WjX1Vk8tWZtjzOz/9Q48rNhAQgQbSIhgAwkRbCAhgg0kVPSp+OBBkGODzc2Sbkq6ONieiYjrDWoDMKaiYEfEJUk7JMn2vKQrEXGwYV0AJsCpOJAQwQYSqhrsla2RLl66WXNoACOoGuyVrZE2blhTc2gAI+BUHEiIYAMJjfx0V0TMN6gDQEWs2EBCBBtIiGADCTXroFJLTx0raqPzyf+WdV63Ays2kBDBBhIi2EBCBBtIiGADCRUF2/Ym2x/Y/t72adsnbM+2Lg7AeIYG27YlfSTps4h4KCJ2SXpJ0gOtiwMwnpLr2HslXY+IQ8s7IuK8pHeaVQVgIiWn4o9J+rp1IQDqGfnDM9vv2j5j+6tVXqODCtCBkmAvStq5vBERr0vaJ2njrW+kgwrQh5JgH5e0zvZrK/atb1QPgAqGBjsiQtIBSXtsn7N9UtJ7kt5sXRyA8ZR+YcCPWrrEBeAOwJ1nQEIEG0iIYAMJeemzsfru9X3xlPdNPE5v3S9qdfWQ+psbhqt5/Cf1ZRzTT3HZq73Gig0kRLCBhAg2kBDBBhIi2EBCBBtIqDjYtmdth+1HWhYEYHKjrNhzkr4Q94wD3SttZniPpKclvSKCDXSvdMU+IOmTiDgr6bLtnau9aWUHlRu6Vq1IAKMpDfacpIXBzwuD7V9Y2UHlLq2tUR+AMQx9Htv2Bi11Kt1uOyStkRS234hWN5oDmEjJiv2CpPcjYmtEPBgRWySdk/RM29IAjKsk2HOSjtyy70NJL9cvB0ANQ0/FI+LZVfa93aQaAFVw5xmQEMEGEiLYQEJF7YfHse3xqzp6dPLWPz21opHqtjOqNbdaNfVWTy1Z21nN7L/6q6+xYgMJEWwgIYINJESwgYQINpBQ0afigwdBjg02N0u6KeniYHsmIq43qA3AmEq/bfOSpB2SZHte0pWIONiwLgAT4FQcSKhqsFd2ULl46WbNoQGMoGqwV3ZQ2bhhTc2hAYyAU3EgIYINJESwgYRGfrorIuYb1AGgIlZsICGCDSREsIGE3Krn/+4n1sXJo1uajI2+Ze3E0puZ/T/o1JmfvdprrNhAQgQbSIhgAwkRbCAhgg0kVBxs25ttL9j+zva3tj+2va1lcQDGUxRs29bSN25+GhEPR8Sjkt6StKllcQDGU3qv+HOSbkTEoeUdEcFFRqBTpafi2yWdHvYmOqgAfaCDCpBQabAXJe1qWQiAekqDfVzSWtuvLu+w/aTtPW3KAjCJomDH0pMis5KeH1zuWpQ0L+lCw9oAjKm4g0pEXJD0YsNaAFTCnWdAQgQbSIhgAwmN3KX0dqvVjaOWml09eus00ls9tdT8Heptbr+GFRtIiGADCRFsICGCDSREsIGEij4Vt71B0rHB5mZJNyVdHGzPRMT1BrUBGFNRsCPikqQdkmR7XtKViDjYsC4AE+BUHEioarDpoAL0gQ4qQEKcigMJEWwgIYINJDTy010RMd+gDgAVsWIDCRFsICGCDSTkpc7C9e1+Yl2cPLqlydjoW9ZOLL2Z2f+DTp352au9xooNJESwgYQINpAQwQYSIthAQsXBtr3Z9sLgS/m+tf2x7W0tiwMwnqJg27akI5I+jYiHI+JRSW9J2tSyOADjKb1X/DlJNyLi0PKOiOAiI9Cp0lPx7ZJOD3sTHVSAPtBBBUioNNiLkna1LARAPaXBPi5pre1Xl3fYftL2njZlAZhEUbBj6UmRWUnPDy53LUqal3ShYW0AxlTcQSUiLkh6sWEtACrhzjMgIYINJESwgYRG7lKK/vTWsaTWOL3N607Cig0kRLCBhAg2kBDBBhIi2EBCpY0WNtn+wPb3tk/bPmF7tnVxAMYzNNiD7ikfSfosIh6KiF2SXpL0QOviAIyn5Dr2XknXb+mecl7SO82qAjCRklPxxyR9XTIYHVSAPoz84Zntd22fsf3Vra/RQQXoQ0mwFyXtXN6IiNcl7ZO0sVVRACZTEuzjktbZfm3FvvWN6gFQwdBgD7qnHJC0x/Y52yclvSfpzdbFARhP0dNdEfGjli5xAbgDcOcZkBDBBhIi2EBC3XdQqdVFo5aa3Th66xDSWz29zUu6c7qxsGIDCRFsICGCDSREsIGECDaQUNGn4rY3SDo22Nws6aaki4PtmYi43qA2AGMqvaX0kqQdkmR7XtKViDjYsC4AE+BUHEiIYAMJVQ02rZGAPlQNNq2RgD5wKg4kRLCBhEZ+uisi5hvUAaAiVmwgIYINJESwgYQINpCQl9qG13ev74unvG/icXprRfP/2GYH/9VTq64v45h+iste7TVWbCAhgg0kRLCBhAg2kFBxsG3P2g7bj7QsCMDkRlmx5yR9Ib6cD+heUbBt3yPpaUmviGAD3StdsQ9I+iQizkq6bHtnw5oATKg02HOSFgY/Lwy2f2FlB5UbulajPgBjGPrY5qD18F5J222HpDWSwvYbccttaxFxWNJhaenOswb1AihQsmK/IOn9iNgaEQ9GxBZJ5yQ907Y0AOMqCfacpCO37PtQ0sv1ywFQw9BT8Yh4dpV9bzepBkAV3HkGJESwgYQINpAQwQYSIthAQgQbSIhgAwkRbCAhgg0kRLCBhGiNBCREayQgIVojAQlVbY1EBxWgD1VbI0XE4YjYHRG779LaGvUBGEPV1kgA+kBrJCAhWiMBCdEaCUiIO8+AhAg2kBDBBhJyqytWti9KOj/kbfdL+leFv45xbt9YjNPPOFsjYuNqLzQLdgnbpyJiN+O0HafHmhin7TicigMJEWwgoWkH+zDj3JZxao7FOHfAOFP9PzaANqa9YgNogGADCU0t2JP2ULO9wfY3gz//tP2PFdt33+5xVozXxbwGY22y/YHt722ftn3C9uyUa9pse8H2d7a/tf2x7W1j1DTx3Ho79lXHiYip/JH0N0mfS5qvMNa8pL/2ME4v85JkSSck/XnFvq2S/tJZTTsk/WHac+vp2NcYZyordtYeap3Na6+k6xFxaHlHRJyPiHemWNNzkm7cUtM3EfH5iON0N7dax77WONM6FS/qoXYH6mlej0n6eop//2q2SzpdYZwe51br2FcZZ1rBLuqhdgfqdl6237V9xvZX066ltk7mVuvYVxlnaKOF2rL2UOtwXouS/ri8ERGv275f0qkp1LJsUUuttmqM083cah37mr9D01ixs/ZQ621exyWts/3ain3rp1TLsuOS1tp+dXmH7Sdt7xljnJ7mVuvYV/sdmkaws/ZQ62peg3/hD0jaY/uc7ZOS3pP05jTqWVHTrKTnB5e7FrX0afSFMcbpaW61jn213yFuKQUS4s4zICGCDSREsIGECDaQEMEGEiLYQEIEG0joP11mpsHmlY7MAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dotplotx(dna1,dna2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPYAAAD4CAYAAADfGBu4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAALiUlEQVR4nO3d0YvVZR7H8c9n3VLc6CITZUmMAomSTdTsohZLiW4dVqLpNlo2ZO+Wgq7mD/CqCMS7uohhIewq6kKJCiTTyIXpQiiRWFtw9SLcJUfkuxdzZneYpj3POed5nGe+vl8gzO93xsfvM7/z8fnN+f3O9zgiBCCXX612AQDqI9hAQgQbSIhgAwkRbCChX7ca+G6vjw36zcTj7PjdvytUU8+Fv22sNlZvc8NwNY//pH7SvzQfN7zSY251uete3xdP+uDE43x8+esK1dTz/G93VRurt7lhuJrHf1JfxEn9GNdWDDan4kBCBBtIiGADCRFsIKHiYNuesh22H2lZEIDJjbJiT0v6XNKLjWoBUElRsG3fI+kpSS+LYAPdK12xD0n6KCIuSLpme/dK32T7j7bP2j57UzeqFQlgNKXBnpY0O/h6drD9MxFxPCL2RsTeu7S+Rn0AxjD0llLbmyQdkLTTdkhaJylsvxZ0aQC6VLJiH5b0bkRsj4gHI2KbpIuSnm5bGoBxlQR7WtKJZfvel/RS/XIA1DD0VDwinllh35tNqgFQBXeeAQkRbCAhgg0k1KzRwt7HN8SZj7dNPE6tN7b32NSgt7n1Vk8tWZtj7Hv+e509/xONFoA7BcEGEiLYQEIEG0iIYAMJFfUVH7wR5ORgc6ukW5KuDLb3RcR8g9oAjKko2BFxVdIuSbI9I+l6RBxtWBeACXAqDiREsIGEqgZ7aWukK1dv1RwawAiqBntpa6TNm9bVHBrACDgVBxIi2EBCI38+dkTMNKgDQEWs2EBCBBtIiGADCY38O/btlrk7SE/dOKT+6unxmK0VrNhAQgQbSIhgAwkRbCAhgg0kVBRs21tsv2f7O9vnbJ+2PdW6OADjGRps25b0gaRPI+KhiNgj6UVJD7QuDsB4Sq5jH5A0HxHHFndExCVJbzWrCsBESk7FH5P0VetCANQz8otntt+2fd72lys8RgcVoAMlwZ6TtHtxIyKOSDooafPyb6SDCtCHkmCfkrTB9qtL9m1sVA+ACoYGOxY+Z/eQpP22L9o+I+kdSa+3Lg7AeEo/MOAHLVziArAGcOcZkBDBBhIi2EBCXnhtrL57fV886YMTj9Nb94taXT2k/uaG4Woe/0l9ESf1Y1zzSo+xYgMJEWwgIYINJESwgYQINpAQwQYSKg627SnbYfuRlgUBmNwoK/a0pM/FPeNA90qbGd4j6SlJL4tgA90rXbEPSfooIi5IumZ790rftLSDyk3dqFYkgNGUBnta0uzg69nB9s8s7aByl9bXqA/AGIa+H9v2Ji10Kt1pOyStkxS2X4tWN5oDmEjJin1Y0rsRsT0iHoyIbZIuSnq6bWkAxlUS7GlJJ5bte1/SS/XLAVDD0FPxiHhmhX1vNqkGQBXceQYkRLCBhAg2kFCz1kh7H98QZz7eNvE4tVrR9NiGqLe59VZPLVnbWe17/nudPf8TrZGAOwXBBhIi2EBCBBtIiGADCRV9KN/gjSAnB5tbJd2SdGWwvS8i5hvUBmBMpZ+2eVXSLkmyPSPpekQcbVgXgAlwKg4kVDXYSzuoXLl6q+bQAEZQNdhLO6hs3rSu5tAARsCpOJAQwQYSIthAQkWXu5aKiJkGdQCoiBUbSIhgAwkRbCChkX/Hvt166lhRW9bOJ73Vk/k59EtYsYGECDaQEMEGEiLYQEIEG0ioONi2t9qetf2t7W9sf2h7R8viAIynKNi2rYVP3PwkIh6OiEclvSFpS8viAIyn9Dr2s5JuRsSxxR0RceddHATWiNJT8Z2Szg37JjqoAH2ggwqQUGmw5yTtaVkIgHpKg31K0nrbryzusP2E7f1tygIwiaJgx8Jn7U5Jem5wuWtO0oykyw1rAzCm4nd3RcRlSS80rAVAJdx5BiREsIGECDaQUPcdVHrrxlFTb3PrrZ5aas1L6m9uv4QVG0iIYAMJEWwgIYINJESwgYSKXhW3vUnSycHmVkm3JF0ZbO+LiPkGtQEYU1GwI+KqpF2SZHtG0vWIONqwLgAT4FQcSKhqsOmgAvSBDipAQpyKAwkRbCAhgg0kNPK7uyJipkEdACpixQYSIthAQgQbSKj7DiprpWPFOLJ2PumtnszPoV/Cig0kRLCBhAg2kBDBBhIi2EBCxcG2vdX27OBD+b6x/aHtHS2LAzCeomDbtqQTkj6JiIcj4lFJb0ja0rI4AOMpvY79rKSbEXFscUdE3HkXB4E1ovRUfKekc8O+iQ4qQB/ooAIkVBrsOUl7WhYCoJ7SYJ+StN72K4s7bD9he3+bsgBMoijYERGSpiQ9N7jcNSdpRtLlhrUBGFPxu7si4rKkFxrWAqAS7jwDEiLYQEIEG0io+w4qtfTW1aOm3mrK2ollLWHFBhIi2EBCBBtIiGADCRFsIKHSRgtbbL9n+zvb52yftj3VujgA4xka7EH3lA8kfRoRD0XEHkkvSnqgdXEAxlNyHfuApPll3VMuSXqrWVUAJlJyKv6YpK9KBqODCtCHkV88s/227fO2v1z+GB1UgD6UBHtO0u7FjYg4IumgpM2tigIwmZJgn5K0wfarS/ZtbFQPgAqGBnvQPeWQpP22L9o+I+kdSa+3Lg7AeIre3RURP2jhEheANYA7z4CECDaQEMEGEuq+g0rmLhq9za23enqbl9Tn82glrNhAQgQbSIhgAwkRbCAhgg0kVPSquO1Nkk4ONrdKuiXpymB7X0TMN6gNwJhKbym9KmmXJNmekXQ9Io42rAvABDgVBxIi2EBCVYNNaySgD1WDTWskoA+cigMJEWwgoZHf3RURMw3qAFARKzaQEMEGEiLYQEIEG0jIC23D67vX98WTPjjxOL21orkT2+zgf2oe/0l9ESf1Y1zzSo+xYgMJEWwgIYINJESwgYSKg217ynbYfqRlQQAmN8qKPS3pc/HhfED3ioJt+x5JT0l6WQQb6F7pin1I0kcRcUHSNdu7G9YEYEKlwZ6WNDv4enaw/TNLO6jc1I0a9QEYw9C3bQ5aDx+QtNN2SFonKWy/FstuW4uI45KOSwt3njWoF0CBkhX7sKR3I2J7RDwYEdskXZT0dNvSAIyrJNjTkk4s2/e+pJfqlwOghqGn4hHxzAr73mxSDYAquPMMSIhgAwkRbCAhgg0kRLCBhAg2kBDBBhIi2EBCBBtIiGADCdEaCUiI1khAQrRGAhKq2hqJDipAH6q2RoqI4xGxNyL23qX1NeoDMIaqrZEA9IHWSEBCtEYCEqI1EpAQd54BCRFsICGCDSTkVlesbF+RdGnIt90v6Z8V/jnGuX1jMU4/42yPiM0rPdAs2CVsn42IvYzTdpwea2KctuNwKg4kRLCBhFY72McZ57aMU3MsxlkD46zq79gA2ljtFRtAAwQbSGjVgj1pDzXbm2x/PfjzD9t/X7J994hjbbH9nu3vbJ+zfdr21Jh1dTOvwXhbbc/a/tb2N7Y/tL1jtWqqUc9gnImPWYOfdZW+gFXGiYhV+SPpr5I+kzRTYawZSX8Z8+9a0mlJf1qyb7ukP6/lef2fue2S9PuOftYj11P7mNX4Wdc89jXGWZUVu7MeagckzUfEscUdEXEpIt4adaDO5iVJz0q6uWxuX0fEZ2u8nmrHrJZax77WOKt1Kl7UQ+02eUzSV5XG6mlekrRT0rlVrmGpWvXUPGa11Dr2VcZZrWAX9VBbDbbftn3e9pdj/PVu55XZhMesllrHvso4Qxst1NZhD7U5SX9Y3IiII7bvl3R2lEE6nJe0MLfDq/Rvr6RWPVWOWS21jn3N59BqrNi99VA7JWmD7VeX7Ns4xji9zUtamNt6268s7rD9hO39a7yeWsesllrHvtpzaDWC3VUPtcH/hIck7bd90fYZSe9Ien3Eobqal/TfuU1Jem5weWlOC6/+Xl7L9VQ8ZrXUOvbVnkPcUgokxJ1nQEIEG0iIYAMJEWwgIYINJESwgYQINpDQfwCMasPnr97vcQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dotplotx(dna1,dna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPYAAAD4CAYAAADfGBu4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAALiUlEQVR4nO3d0YvVZR7H8c9n3VLc6CITZUmMAomSTdTsohZLiW4dVqLpNlo2ZO+Wgq7mD/CqCMS7uohhIewq6kKJCiTTyIXpQiiRWFtw9SLcJUfkuxdzZneYpj3POed5nGe+vl8gzO93xsfvM7/z8fnN+f3O9zgiBCCXX612AQDqI9hAQgQbSIhgAwkRbCChX7ca+G6vjw36zcTj7PjdvytUU8+Fv22sNlZvc8NwNY//pH7SvzQfN7zSY251uete3xdP+uDE43x8+esK1dTz/G93VRurt7lhuJrHf1JfxEn9GNdWDDan4kBCBBtIiGADCRFsIKHiYNuesh22H2lZEIDJjbJiT0v6XNKLjWoBUElRsG3fI+kpSS+LYAPdK12xD0n6KCIuSLpme/dK32T7j7bP2j57UzeqFQlgNKXBnpY0O/h6drD9MxFxPCL2RsTeu7S+Rn0AxjD0llLbmyQdkLTTdkhaJylsvxZ0aQC6VLJiH5b0bkRsj4gHI2KbpIuSnm5bGoBxlQR7WtKJZfvel/RS/XIA1DD0VDwinllh35tNqgFQBXeeAQkRbCAhgg0k1KzRwt7HN8SZj7dNPE6tN7b32NSgt7n1Vk8tWZtj7Hv+e509/xONFoA7BcEGEiLYQEIEG0iIYAMJFfUVH7wR5ORgc6ukW5KuDLb3RcR8g9oAjKko2BFxVdIuSbI9I+l6RBxtWBeACXAqDiREsIGEqgZ7aWukK1dv1RwawAiqBntpa6TNm9bVHBrACDgVBxIi2EBCI38+dkTMNKgDQEWs2EBCBBtIiGADCY38O/btlrk7SE/dOKT+6unxmK0VrNhAQgQbSIhgAwkRbCAhgg0kVBRs21tsv2f7O9vnbJ+2PdW6OADjGRps25b0gaRPI+KhiNgj6UVJD7QuDsB4Sq5jH5A0HxHHFndExCVJbzWrCsBESk7FH5P0VetCANQz8otntt+2fd72lys8RgcVoAMlwZ6TtHtxIyKOSDooafPyb6SDCtCHkmCfkrTB9qtL9m1sVA+ACoYGOxY+Z/eQpP22L9o+I+kdSa+3Lg7AeEo/MOAHLVziArAGcOcZkBDBBhIi2EBCXnhtrL57fV886YMTj9Nb94taXT2k/uaG4Woe/0l9ESf1Y1zzSo+xYgMJEWwgIYINJESwgYQINpAQwQYSKg627SnbYfuRlgUBmNwoK/a0pM/FPeNA90qbGd4j6SlJL4tgA90rXbEPSfooIi5IumZ790rftLSDyk3dqFYkgNGUBnta0uzg69nB9s8s7aByl9bXqA/AGIa+H9v2Ji10Kt1pOyStkxS2X4tWN5oDmEjJin1Y0rsRsT0iHoyIbZIuSnq6bWkAxlUS7GlJJ5bte1/SS/XLAVDD0FPxiHhmhX1vNqkGQBXceQYkRLCBhAg2kFCz1kh7H98QZz7eNvE4tVrR9NiGqLe59VZPLVnbWe17/nudPf8TrZGAOwXBBhIi2EBCBBtIiGADCRV9KN/gjSAnB5tbJd2SdGWwvS8i5hvUBmBMpZ+2eVXSLkmyPSPpekQcbVgXgAlwKg4kVDXYSzuoXLl6q+bQAEZQNdhLO6hs3rSu5tAARsCpOJAQwQYSIthAQkWXu5aKiJkGdQCoiBUbSIhgAwkRbCChkX/Hvt166lhRW9bOJ73Vk/k59EtYsYGECDaQEMEGEiLYQEIEG0ioONi2t9qetf2t7W9sf2h7R8viAIynKNi2rYVP3PwkIh6OiEclvSFpS8viAIyn9Dr2s5JuRsSxxR0RceddHATWiNJT8Z2Szg37JjqoAH2ggwqQUGmw5yTtaVkIgHpKg31K0nrbryzusP2E7f1tygIwiaJgx8Jn7U5Jem5wuWtO0oykyw1rAzCm4nd3RcRlSS80rAVAJdx5BiREsIGECDaQUPcdVHrrxlFTb3PrrZ5aas1L6m9uv4QVG0iIYAMJEWwgIYINJESwgYSKXhW3vUnSycHmVkm3JF0ZbO+LiPkGtQEYU1GwI+KqpF2SZHtG0vWIONqwLgAT4FQcSKhqsOmgAvSBDipAQpyKAwkRbCAhgg0kNPK7uyJipkEdACpixQYSIthAQgQbSKj7DiprpWPFOLJ2PumtnszPoV/Cig0kRLCBhAg2kBDBBhIi2EBCxcG2vdX27OBD+b6x/aHtHS2LAzCeomDbtqQTkj6JiIcj4lFJb0ja0rI4AOMpvY79rKSbEXFscUdE3HkXB4E1ovRUfKekc8O+iQ4qQB/ooAIkVBrsOUl7WhYCoJ7SYJ+StN72K4s7bD9he3+bsgBMoijYERGSpiQ9N7jcNSdpRtLlhrUBGFPxu7si4rKkFxrWAqAS7jwDEiLYQEIEG0io+w4qtfTW1aOm3mrK2ollLWHFBhIi2EBCBBtIiGADCRFsIKHSRgtbbL9n+zvb52yftj3VujgA4xka7EH3lA8kfRoRD0XEHkkvSnqgdXEAxlNyHfuApPll3VMuSXqrWVUAJlJyKv6YpK9KBqODCtCHkV88s/227fO2v1z+GB1UgD6UBHtO0u7FjYg4IumgpM2tigIwmZJgn5K0wfarS/ZtbFQPgAqGBnvQPeWQpP22L9o+I+kdSa+3Lg7AeIre3RURP2jhEheANYA7z4CECDaQEMEGEuq+g0rmLhq9za23enqbl9Tn82glrNhAQgQbSIhgAwkRbCAhgg0kVPSquO1Nkk4ONrdKuiXpymB7X0TMN6gNwJhKbym9KmmXJNmekXQ9Io42rAvABDgVBxIi2EBCVYNNaySgD1WDTWskoA+cigMJEWwgoZHf3RURMw3qAFARKzaQEMEGEiLYQEIEG0jIC23D67vX98WTPjjxOL21orkT2+zgf2oe/0l9ESf1Y1zzSo+xYgMJEWwgIYINJESwgYSKg217ynbYfqRlQQAmN8qKPS3pc/HhfED3ioJt+x5JT0l6WQQb6F7pin1I0kcRcUHSNdu7G9YEYEKlwZ6WNDv4enaw/TNLO6jc1I0a9QEYw9C3bQ5aDx+QtNN2SFonKWy/FstuW4uI45KOSwt3njWoF0CBkhX7sKR3I2J7RDwYEdskXZT0dNvSAIyrJNjTkk4s2/e+pJfqlwOghqGn4hHxzAr73mxSDYAquPMMSIhgAwkRbCAhgg0kRLCBhAg2kBDBBhIi2EBCBBtIiGADCdEaCUiI1khAQrRGAhKq2hqJDipAH6q2RoqI4xGxNyL23qX1NeoDMIaqrZEA9IHWSEBCtEYCEqI1EpAQd54BCRFsICGCDSTkVlesbF+RdGnIt90v6Z8V/jnGuX1jMU4/42yPiM0rPdAs2CVsn42IvYzTdpwea2KctuNwKg4kRLCBhFY72McZ57aMU3MsxlkD46zq79gA2ljtFRtAAwQbSGjVgj1pDzXbm2x/PfjzD9t/X7J994hjbbH9nu3vbJ+zfdr21Jh1dTOvwXhbbc/a/tb2N7Y/tL1jtWqqUc9gnImPWYOfdZW+gFXGiYhV+SPpr5I+kzRTYawZSX8Z8+9a0mlJf1qyb7ukP6/lef2fue2S9PuOftYj11P7mNX4Wdc89jXGWZUVu7MeagckzUfEscUdEXEpIt4adaDO5iVJz0q6uWxuX0fEZ2u8nmrHrJZax77WOKt1Kl7UQ+02eUzSV5XG6mlekrRT0rlVrmGpWvXUPGa11Dr2VcZZrWAX9VBbDbbftn3e9pdj/PVu55XZhMesllrHvso4Qxst1NZhD7U5SX9Y3IiII7bvl3R2lEE6nJe0MLfDq/Rvr6RWPVWOWS21jn3N59BqrNi99VA7JWmD7VeX7Ns4xji9zUtamNt6268s7rD9hO39a7yeWsesllrHvtpzaDWC3VUPtcH/hIck7bd90fYZSe9Ien3Eobqal/TfuU1Jem5weWlOC6/+Xl7L9VQ8ZrXUOvbVnkPcUgokxJ1nQEIEG0iIYAMJEWwgIYINJESwgYQINpDQfwCMasPnr97vcQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dna1 = Seq(\"ATGATCTCGTAA\",generic_dna)\n",
    "dna2 = Seq(\"ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAA\",generic_dna)\n",
    "\n",
    "dotplotx(dna1,dna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check for the Count of Amino Acids and Molecular Weight\n",
    "+ ProtParam.ProteinAnalysis\n",
    "+ Counter from collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import molecular_weight\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3724.3894999999998"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna = Seq('ATGATCTCGTAA')\n",
    "\n",
    "# Moleculer weight of DNA nucleotide\n",
    "molecular_weight(dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Protein\n",
    "protein = dna.translate()\n",
    "protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amino Acid Count\n",
    "protein_a = ProteinAnalysis(str(protein))\n",
    "\n",
    "# dir(protein_a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 0,\n",
       " 'C': 0,\n",
       " 'D': 0,\n",
       " 'E': 0,\n",
       " 'F': 0,\n",
       " 'G': 0,\n",
       " 'H': 0,\n",
       " 'I': 1,\n",
       " 'K': 0,\n",
       " 'L': 0,\n",
       " 'M': 1,\n",
       " 'N': 0,\n",
       " 'P': 0,\n",
       " 'Q': 0,\n",
       " 'R': 0,\n",
       " 'S': 1,\n",
       " 'T': 0,\n",
       " 'V': 0,\n",
       " 'W': 0,\n",
       " 'Y': 0}"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count of AA\n",
    "protein_a.count_amino_acids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method 2 \n",
    "# Counter\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'M': 1, 'I': 1, 'S': 1, '*': 1})"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_freq = Counter(protein)\n",
    "aa_freq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 4 artists>"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAMCklEQVR4nO3df6jd913H8edribHC5gbmCjM/loKZGMqw81In+2OVbZB2kvwzJNGhk7L8Y/WPDUdkUkv9RzthqGRqwFLtsFndH3qZGRloiyK25JbNzrREL3FbrhGabrVQ51qjb/+4p/Ps9tyc78096e195/mAkPP9fj85953vH0+++Z77vUlVIUna+t6w2QNIkmbDoEtSEwZdkpow6JLUhEGXpCa2b9YX3rlzZ+3bt2+zvrwkbUlPPvnkc1U1N+nYpgV93759LC4ubtaXl6QtKcnX1jrmLRdJasKgS1ITBl2SmjDoktSEQZekJgy6JDUxNehJHkjybJJ/WuN4kvxekqUkTyV55+zHlCRNM+QK/UHg4FWO3wHsH/06BvzBxseSJK3X1KBX1d8C37zKksPAn9aKx4G3JHnrrAaUJA0ziydFdwEXx7aXR/v+ffXCJMdYuYpn79691/wF9x3/q2v+sx189bc+sKE/7/nz/G2U53BjNnr+1jKLD0UzYd/E/wapqk5W1XxVzc/NTfxRBJKkazSLoC8De8a2dwOXZvC+kqR1mEXQF4CfH323y7uAF6rqVbdbJEnX19R76EkeBm4HdiZZBn4D+B6AqvpD4DRwJ7AEfAv4xes1rCRpbVODXlVHpxwv4JdmNpEk6Zr4pKgkNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYGBT3JwSTnkywlOT7h+N4kjyb5UpKnktw5+1ElSVczNehJtgEngDuAA8DRJAdWLft14JGquhU4Anx61oNKkq5uyBX6bcBSVV2oqpeBU8DhVWsK+P7R6zcDl2Y3oiRpiCFB3wVcHNteHu0bdy/woSTLwGnglye9UZJjSRaTLF6+fPkaxpUkrWVI0DNhX63aPgo8WFW7gTuBh5K86r2r6mRVzVfV/Nzc3PqnlSStaUjQl4E9Y9u7efUtlbuARwCq6h+Am4CdsxhQkjTMkKCfBfYnuTnJDlY+9FxYtebrwHsBkvwoK0H3nookvYamBr2qrgB3A2eAZ1j5bpZzSe5Lcmi07GPAR5L8I/Aw8OGqWn1bRpJ0HW0fsqiqTrPyYef4vnvGXj8NvHu2o0mS1sMnRSWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNTEo6EkOJjmfZCnJ8TXW/EySp5OcS/Jnsx1TkjTN9mkLkmwDTgDvB5aBs0kWqurpsTX7gV8D3l1Vzyf5wes1sCRpsiFX6LcBS1V1oapeBk4Bh1et+QhwoqqeB6iqZ2c7piRpmiFB3wVcHNteHu0b93bg7Un+PsnjSQ7OakBJ0jBTb7kAmbCvJrzPfuB2YDfwd0luqar/+K43So4BxwD27t277mElSWsbcoW+DOwZ294NXJqw5i+r6r+r6l+B86wE/rtU1cmqmq+q+bm5uWudWZI0wZCgnwX2J7k5yQ7gCLCwas1fAD8FkGQnK7dgLsxyUEnS1U0NelVdAe4GzgDPAI9U1bkk9yU5NFp2BvhGkqeBR4FfrapvXK+hJUmvNuQeOlV1Gji9at89Y68L+OjolyRpE/ikqCQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDUxKOhJDiY5n2QpyfGrrPtgkkoyP7sRJUlDTA16km3ACeAO4ABwNMmBCeveBPwK8MSsh5QkTTfkCv02YKmqLlTVy8Ap4PCEdb8J3A98e4bzSZIGGhL0XcDFse3l0b7vSHIrsKeqPn+1N0pyLMliksXLly+ve1hJ0tqGBD0T9tV3DiZvAD4FfGzaG1XVyaqar6r5ubm54VNKkqYaEvRlYM/Y9m7g0tj2m4BbgMeSfBV4F7DgB6OS9NoaEvSzwP4kNyfZARwBFl45WFUvVNXOqtpXVfuAx4FDVbV4XSaWJE00NehVdQW4GzgDPAM8UlXnktyX5ND1HlCSNMz2IYuq6jRwetW+e9ZYe/vGx5IkrZdPikpSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJamJQ0JMcTHI+yVKS4xOOfzTJ00meSvLXSd42+1ElSVczNehJtgEngDuAA8DRJAdWLfsSMF9V7wA+B9w/60ElSVc35Ar9NmCpqi5U1cvAKeDw+IKqerSqvjXafBzYPdsxJUnTDAn6LuDi2PbyaN9a7gK+MOlAkmNJFpMsXr58efiUkqSphgQ9E/bVxIXJh4B54JOTjlfVyaqar6r5ubm54VNKkqbaPmDNMrBnbHs3cGn1oiTvAz4BvKeqXprNeJKkoYZcoZ8F9ie5OckO4AiwML4gya3AHwGHqurZ2Y8pSZpmatCr6gpwN3AGeAZ4pKrOJbkvyaHRsk8CbwT+PMmXkyys8XaSpOtkyC0Xquo0cHrVvnvGXr9vxnNJktbJJ0UlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpiUFBT3IwyfkkS0mOTzj+vUk+Ozr+RJJ9sx5UknR1U4OeZBtwArgDOAAcTXJg1bK7gOer6oeBTwG/PetBJUlXN+QK/TZgqaouVNXLwCng8Ko1h4E/Gb3+HPDeJJndmJKkabYPWLMLuDi2vQz8xFprqupKkheAHwCeG1+U5BhwbLT5YpLz1zL068BOVv3dXkvZ+v/+8fxtnOdwY7by+XvbWgeGBH3SlXZdwxqq6iRwcsDXfF1LslhV85s9x1bl+ds4z+HGdD1/Q265LAN7xrZ3A5fWWpNkO/Bm4JuzGFCSNMyQoJ8F9ie5OckO4AiwsGrNAvALo9cfBP6mql51hS5Jun6m3nIZ3RO/GzgDbAMeqKpzSe4DFqtqAfhj4KEkS6xcmR+5nkO/Dmz520abzPO3cZ7DjWl5/uKFtCT14JOiktSEQZekJgz6QEkqyUNj29uTXE7y+c2caytK8uJmz7BVJflEknNJnkry5SSrnwnRGl552DHJvePbnQz5PnSt+E/gliTfV1X/Bbwf+LdNnkk3kCQ/Cfw08M6qeinJTmDHJo+1lfxckh8CbkrycVa+/fozmzzTTHmFvj5fAD4wen0UeHgTZ9GN563Ac1X1EkBVPVdVq58J0Rqq6jOsPNH+ceDro+1WDPr6nAKOJLkJeAfwxCbPoxvLF4E9Sf45yaeTvGezB9pKkvwsKw9A3g/sHW23YtDXoaqeAvaxcnV+enOn0Y2mql4EfpyVn4d0Gfhskg9v6lBby8NVdT/w7dHv7f6F7T309VsAfge4nZUfQCa9Zqrqf4DHgMeSfIWVJ7Qf3MyZtopXnl6vqnvHtzsx6Ov3APBCVX0lye2bPYxuHEl+BPjfqvqX0a4fA762iSPpdcagr1NVLQO/u9lz6Ib0RuD3k7wFuAIs8f8/jlry0X9J6sIPRSWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6Qm/g9r1aop8HpaigAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.bar(aa_freq.keys(), aa_freq.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "349.4462"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p2 = Seq('MIS')\n",
    "\n",
    "protein2 = ProteinAnalysis(str(p2))\n",
    "protein2.molecular_weight()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.275022315979004"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein2.isoelectric_point()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with File Formats\n",
    " + FASTA\n",
    " + GenBank\n",
    " + PDB\n",
    " + etc\n",
    "\n",
    "+ Sample https://www.ncbi.nlm.nih.gov/nuccore/MN908947"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MN908947.3\n",
      "MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "-------------------------------\n",
      "ID: MN908947.3\n",
      "Name: MN908947.3\n",
      "Description: MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 0\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "# Reading FASTA\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Load a FASTA\n",
    "for record in SeqIO.parse(\"sequence.fasta\",\"fasta\"):\n",
    "    print(record.id)\n",
    "    print(record.description)\n",
    "    print(\"-------------------------------\")\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', SingleLetterAlphabet())"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading the Sequence in FASTA\n",
    "dna_record = SeqIO.read(\"sequence.fasta\",\"fasta\")\n",
    "dna_seq = dna_record.seq\n",
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: MN908947.3\n",
      "Name: MN908947\n",
      "Description: Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 23\n",
      "/molecule_type=ss-RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=18-MAR-2020\n",
      "/accessions=['MN908947']\n",
      "/sequence_version=3\n",
      "/keywords=['']\n",
      "/source=Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)\n",
      "/organism=Severe acute respiratory syndrome coronavirus 2\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='A new coronavirus associated with human respiratory disease in China', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=On Jan 17, 2020 this sequence version replaced MN908947.2.\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'Megahit v. V1.1.3'), ('Sequencing Technology', 'Illumina')]))])\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', IUPACAmbiguousDNA())\n"
     ]
    }
   ],
   "source": [
    "# Reading GenBank\n",
    "# Load a GenBank File\n",
    "for record in SeqIO.parse(\"sequence.gb\",\"genbank\"):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', IUPACAmbiguousDNA()), id='MN908947.3', name='MN908947', description='Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome', dbxrefs=[])"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading the sequence in GenBank\n",
    "gb_dna_record = SeqIO.read(\"sequence.gb\",\"genbank\")\n",
    "gb_dna_record\n",
    "#gb_dna_record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writing Data\n",
    "from Bio import SeqIO\n",
    "\n",
    "record_dict = SeqIO.to_dict(SeqIO.parse(\"sequence.fasta\",\"fasta\"))\n",
    "\n",
    "with open('output_fasta.fasta', 'w') as handle:\n",
    "        SeqIO.write(record_dict.values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
